src/backend/dx11/src/device.rs

```diff
@@ -1,983 +1,1322 @@
-use std::{cmp, ptr, slice};
-use std::collections::BTreeMap as Map;
-use std::os::raw::c_void;
-use std::sync::Arc;
-use winapi;
-use core::{self, device as d, buffer, texture, mapping};
-use core::memory::{self, Bind, Typed};
-use core::handle::{self as h, Producer};
-use {Resources as R, Share, Buffer, Fence, Texture, Pipeline, Program, Shader};
-use command::RawCommandBuffer;
-use {CommandList, DeferredContext, ShaderModel};
-use native;
-use wio::com::ComPtr;
-
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct MappingGate {
-    pointer: *mut c_void,
-}
+use hal;
+use hal::queue::QueueFamilyId;
+use hal::range::RangeArg;
+use hal::{buffer, device, error, format, image, mapping, memory, pass, pool, pso, query};
 
-unsafe impl Send for MappingGate {}
-unsafe impl Sync for MappingGate {}
+use winapi::shared::dxgi::{IDXGISwapChain, DXGI_SWAP_CHAIN_DESC, DXGI_SWAP_EFFECT_DISCARD};
+use winapi::shared::minwindef::{FALSE, TRUE};
+use winapi::shared::{dxgiformat, dxgitype, winerror};
+use winapi::um::{d3d11, d3dcommon};
 
-impl core::mapping::Gate<R> for MappingGate {
-    unsafe fn set<T>(&self, index: usize, val: T) {
-        *(self.pointer as *mut T).offset(index as isize) = val;
-    }
+use wio::com::ComPtr;
 
-    unsafe fn slice<'a, 'b, T>(&'a self, len: usize) -> &'b [T] {
-        slice::from_raw_parts(self.pointer as *const T, len)
-    }
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::mem;
+use std::ops::Range;
+use std::ptr;
+use std::sync::Arc;
 
-    unsafe fn mut_slice<'a, 'b, T>(&'a self, len: usize) -> &'b mut [T] {
-        slice::from_raw_parts_mut(self.pointer as *mut T, len)
-    }
-}
+use {
+    Backend, Buffer, BufferView, CommandPool, ComputePipeline, DescriptorPool, DescriptorSetLayout,
+    Fence, Framebuffer, GraphicsPipeline, Image, ImageView, InternalBuffer, Memory, PipelineLayout,
+    QueryPool, RenderPass, Sampler, Semaphore, ShaderModule, Surface, Swapchain, UnboundBuffer,
+    UnboundImage, ViewInfo,
+};
 
-#[derive(Debug)]
-struct TextureParam {
-    levels: winapi::UINT,
-    format: winapi::DXGI_FORMAT,
-    bytes_per_texel: winapi::UINT,
-    bind: winapi::D3D11_BIND_FLAG,
-    usage: winapi::D3D11_USAGE,
-    cpu_access: winapi::D3D11_CPU_ACCESS_FLAG,
-}
+use {conv, internal, shader};
 
 pub struct Device {
-    device: ComPtr<winapi::ID3D11Device>,
-    share: Arc<Share>,
-    frame_handles: h::Manager<R>,
-    vs_cache: Map<u64, Vec<u8>>,
-    /// Create typed surface formats for the textures. This is useful for debugging
-    /// with PIX, since it doesn't understand typeless formats. This may also prevent
-    /// some valid views to be created because the typed formats can't be reinterpret.
-    use_texture_format_hint: bool,
-    sub_data_array: Vec<winapi::D3D11_SUBRESOURCE_DATA>,
-    feature_level: winapi::D3D_FEATURE_LEVEL,
+    pub(crate) device: ComPtr<d3d11::ID3D11Device>,
```

Comment:
nit: I typically name the internal things as `raw` in order to avoid usage like `device.device.xxx`